include/quicly/cc.h

```diff
@@ -33,20 +33,52 @@ extern "C" {
 #include <assert.h>
 #include <stdint.h>
 #include <string.h>
+#include <time.h>
 #include "quicly/constants.h"
+#include "quicly/loss.h"
 
 typedef enum {
     /**
      * Reno, with 0.7 beta reduction
      */
-    CC_RENO_MODIFIED
+    CC_RENO_MODIFIED,
+    /**
+     * CUBIC (RFC 8312)
+     */
+    CC_CUBIC
 } quicly_cc_type_t;
 
+typedef struct st_quicly_cc_conf_t {
+    /**
+     * Congestion controller type.
+     */
+    quicly_cc_type_t type;
+} quicly_cc_conf_t;
```

Comment:
Currently, the `quicly_cc_conf_t` struct indeed serves little purpose. I added it to allow for more configuration options in `quicly_context_t` later, like setting Renos/CUBICs beta value.

You are suggesting to pass a `quicly_cc_impl_t` pointer in `quicly_context_t`, right? That would indeed eliminate the translation, but makes similar configuration parameters harder to implement for users. If they were included in impl, one would have to create a new struct (that has to live as long as the connection) to set those parameters.

What do you think about keeping `quicly_cc_conf_t`, but replacing `quicly_cc_type_t` with a `quicly_cc_impl_t` pointer? Then users can even supply their own CC implementation. type in `quicly_cc_t` would be moved into `quicly_cc_impl_t`. And configuration parameters could still be added later.

Reply:
I do not have a strong opinion if we need to do everything now in this PR, but I would assume that we'd eventually end up in using factory method pattern. As you correctly point out, it is beneficial to have a way to provide CC the configuration parameters. At the same time, each CC might require different set of parameters.

That natural outcome of these requirements is IMO the use of a factory method. The *public* API would look like this:

```c
typedef struct st_quicly_conn_t {
    ...
    /**
     * Creates a congestion controller for given connection.
     */
    quicly_cc_t (*init_cc)(quicly_conn_t *conn);
    ...
} quicly_conn_t;

void quicly_cc_reno_init(quicly_cc_t *cc, uint32_t initcwnd);
void quicly_cc_cubic_init(quicly_cc_t *cc, uint32_t initcwnd, ... /* add more parameters as we need */);
```

There would no longer be a `cc_init` callback in `quicly_cc_impl_t`.

Each user is expected to supply their own `init_cc` callback (we can have defaults associated to the default contexts), that calls the CC they want to use, with the parameters they want to use.

Reply:
I like that! As for doing everything now: especially when it comes to (public) configuration options, I prefer to avoid foreseeable breaking changes.

Will take a shot at implementing this tomorrow. Since `quicly_conn_t` isn't publicly exposed, I'd suggest to pass `quicly_cc_t *` (to modify) + `const quicly_context_t *` (to calculate initcwnd, ...) instead. Not sure what the return value would be used for?

Reply:
Implemented now. Used the existing `QUICLY_CALLBACK_TYPE` macro for init_cc to allow users to pass additional data. The defaults are now located in `default.h/.c` (like the other default callback structs).